xlsx_reader.py

```diff
@@ -0,0 +1,185 @@
+                                   'user': user}
+                book_list.append(book_properties)
+
+            else:
+                asset = current_sheet.cell_value(row_index, 3)
+                book_properties = {'authors': author,
+                                   'current_shelf': current_shelf,
+                                   'title': title,
+                                   'asset': asset,
+                                   'user': user,
+                                   'date_of_rental': date_of_rental,
+                                   'status': status}
+                book_list.append(book_properties)
+
+    return book_list
+
+
+# reading magazine's data from file
+def get_magazine_data(file_location):
+    workbook = load_file(file_location)
+    magazines_list = []
+    current_sheet = workbook.sheet_by_index(2)
+    rows = current_sheet.nrows
+
+    # reading rows except the title of the column
+    for row_index in range(1, rows):
+        title = (current_sheet.cell_value(row_index, 1)).strip()
+        year = current_sheet.cell_value(row_index, 2)
+        issue = current_sheet.cell_value(row_index, 3)
+        magazine_properties = {'title': title, 'year': year, 'issue': issue}
+        magazines_list.append(magazine_properties)
+    return magazines_list
+
+
+# writing authors, books and copies data in database
+def get_books(file_location):
+    books_properties = get_book_data(file_location)
+    asset_codes = []
+
+    for book in books_properties:
+        title = book['title']
+        asset = book['asset']
+        asset_codes.append(asset)
+        authors = book['authors']
+        list_of_authors = []
+
+        if isinstance(authors, tuple):
+            authors_id = []
+            first_name = str(authors[0])
+            last_name = str(authors[1])
+            author = create_library_item(db.session, Author,
+                                         last_name=last_name,
+                                         first_name=first_name)
+            id_of_auth = author.id
+            authors_id.append(id_of_auth)
+            list_of_authors.append(author)
+            book = create_library_item(db.session, Book, title=title)
+            book.authors.append(author)
+
+        elif isinstance(authors, list):
+            authors_id = []
+            for auth_name in authors:
+                f_name = str(auth_name[0])
+                l_name = str(auth_name[1])
+                author = create_library_item(db.session, Author,
+                                             last_name=l_name,
+                                             first_name=f_name)
+                id_of_auth = author.id
+                authors_id.append(id_of_auth)
+                list_of_authors.append(author)
+                book = create_library_item(db.session, Book,
+                                           title=title)
+                book.authors.append(author)
+                if asset in asset_codes:
```

Comment:
I think this "if" statement is always True, because "asset" is added to "asset_codes" on the beginning of outer "for" loop (line 133). Maybe instead this "if" there should be just:

create_library_item(db.session, Copy,
                                        library_item_id=book.id,
                                        library_item=book,
                                        asset_code=asset)

Reply:
But not all books in our current data have asset code, so this is necessary, imho.

Reply:
Thank you for your vigilance, unfortunately asset codes are not provided for every book we have in our file. Deleting this statement would lead to conflicts - according to the db model we are obliged to provide unique asset codes and multiple "None" value violates this rule.